File: ios/HoneycombConfigure.swift
import Honeycomb

@objc public class HoneycombConfigure: NSObject {
    private let builder: HoneycombOptions.Builder
    override internal init() {
        builder = HoneycombOptions.Builder()

            // This produces a lot of extra UI events that are not particularly helpful
            .setUIKitInstrumentationEnabled(false)

            // This locks up react native.
            .setURLSessionInstrumentationEnabled(false)
    }

    @objc public func setAPIKey(_ apiKey: String) -> HoneycombConfigure {
        builder.setAPIKey(apiKey)

        return self
    }

    @objc public func setTracesAPIKey(_ apiKey: String) -> HoneycombConfigure {
        builder.setTracesAPIKey(apiKey)
        return self
    }

    @objc public func setMetricsAPIKey(_ apiKey: String) -> HoneycombConfigure {
        builder.setMetricsAPIKey(apiKey)
        return self
    }

    @objc public func setLogsAPIKey(_ apiKey: String) -> HoneycombConfigure {
        builder.setLogsAPIKey(apiKey)
        return self
    }

    @objc public func setDataset(_ dataset: String) -> HoneycombConfigure {
        builder.setDataset(dataset)
        return self
    }

    @objc public func setMetricsDataset(_ dataset: String) -> HoneycombConfigure {
        builder.setMetricsDataset(dataset)
        return self
    }

    @objc public func setAPIEndpoint(_ endpoint: String) -> HoneycombConfigure {
        builder.setAPIEndpoint(endpoint)
        return self
    }

    @objc public func setTracesAPIEndpoint(_ endpoint: String) -> HoneycombConfigure {
        builder.setTracesAPIEndpoint(endpoint)
        return self
    }

    @objc public func setMetricsAPIEndpoint(_ endpoint: String) -> HoneycombConfigure {
        builder.setMetricsAPIEndpoint(endpoint)
        return self
    }

    @objc public func setLogsAPIEndpoint(_ endpoint: String) -> HoneycombConfigure {
        builder.setLogsAPIEndpoint(endpoint)
        return self
    }

    @objc public func setSampleRate(_ sampleRate: Int) -> HoneycombConfigure {
        builder.setSampleRate(sampleRate)
        return self
    }

    @objc public func setDebug(_ debug: Bool) -> HoneycombConfigure {
        builder.setDebug(debug)
        return self
    }

    @objc public func setServiceName(_ serviceName: String) -> HoneycombConfigure {
        builder.setServiceName(serviceName)
        return self
    }

    @objc public func setServiceVersion(_ serviceVersion: String) -> HoneycombConfigure {
        builder.setServiceVersion(serviceVersion)
        return self
    }

    @objc public func setResourceAttributes(_ resources: [String: String]) -> HoneycombConfigure {
        builder.setResourceAttributes(resources)

        return self
    }

    @objc public func setHeaders(_ headers: [String: String]) -> HoneycombConfigure {
        builder.setHeaders(headers)
        return self
    }

    @objc public func setTracesHeaders(_ headers: [String: String]) -> HoneycombConfigure {
        builder.setTracesHeaders(headers)
        return self
    }

    @objc public func setMetricsHeaders(_ headers: [String: String]) -> HoneycombConfigure {
        builder.setMetricsHeaders(headers)
        return self
    }

    @objc public func setLogsHeaders(_ headers: [String: String]) -> HoneycombConfigure {
        builder.setLogsHeaders(headers)
        return self
    }

    @objc public func setTimeout(_ timeout: TimeInterval) -> HoneycombConfigure {
        builder.setTimeout(timeout)
        return self
    }

    @objc public func setTracesTimeout(_ timeout: TimeInterval) -> HoneycombConfigure {
        builder.setTracesTimeout(timeout)
        return self
    }

    @objc public func setMetricsTimeout(_ timeout: TimeInterval) -> HoneycombConfigure {
        builder.setMetricsTimeout(timeout)
        return self
    }

    @objc public func setLogsTimeout(_ timeout: TimeInterval) -> HoneycombConfigure {
        builder.setLogsTimeout(timeout)
        return self
    }

    @objc public func setSessionTimeout(_ timeout: TimeInterval) -> HoneycombConfigure {
        builder.setSessionTimeout(timeout)
        return self
    }

    @objc public func setMetricKitInstrumentationEnabled(_ enabled: Bool) -> HoneycombConfigure {
        builder.setMetricKitInstrumentationEnabled(enabled)
        return self
    }
    private func setURLSessionInstrumentationEnabled(_ enabled: Bool) -> HoneycombConfigure {
        builder.setURLSessionInstrumentationEnabled(enabled)
        return self
    }
    private func setUIKitInstrumentationEnabled(_ enabled: Bool) -> HoneycombConfigure {
        builder.setUIKitInstrumentationEnabled(enabled)
        return self
    }
    @objc public func setTouchInstrumentationEnabled(_ enabled: Bool) -> HoneycombConfigure {
        builder.setTouchInstrumentationEnabled(enabled)
        return self
    }
    @objc public func setUnhandledExceptionInstrumentationEnabled(_ enabled: Bool)
        -> HoneycombConfigure
    {
        builder.setUnhandledExceptionInstrumentationEnabled(enabled)
        return self
    }

    @objc public func setOfflineCachingEnabled(_ enabled: Bool) -> HoneycombConfigure {
        builder.setOfflineCachingEnabled(enabled)
        return self
    }

    internal func build() throws -> HoneycombOptions {
        return try builder.build()
    }

}

File: android/src/main/java/com/honeycombopentelemetryreactnative/HoneycombConfigure.kt
package com.honeycombopentelemetryreactnative

import android.app.Application
import io.honeycomb.opentelemetry.android.HoneycombOptions
import io.honeycomb.opentelemetry.android.OtlpProtocol
import io.opentelemetry.sdk.logs.LogRecordProcessor
import io.opentelemetry.sdk.trace.SpanProcessor
import kotlin.time.Duration

class HoneycombConfigure(private val context: Application) {
    private val builder : HoneycombOptions.Builder

    init {
        builder = HoneycombOptions.builder(context)
    }

    fun setApiKey(apiKey: String): HoneycombConfigure {
        builder.setApiKey(apiKey)
        return this
    }

    fun setTracesApiKey(apiKey: String): HoneycombConfigure {
        builder.setTracesApiKey(apiKey)
        return this
    }

    fun setMetricsApiKey(apiKey: String): HoneycombConfigure {
        builder.setMetricsApiKey(apiKey)
        return this
    }

    fun setLogsApiKey(apiKey: String): HoneycombConfigure {
        builder.setLogsApiKey(apiKey)
        return this
    }

    fun setDataset(dataset: String): HoneycombConfigure {
        builder.setDataset(dataset)
        return this
    }

    fun setMetricsDataset(dataset: String): HoneycombConfigure {
        builder.setMetricsDataset(dataset)
        return this
    }

    fun setApiEndpoint(endpoint: String): HoneycombConfigure {
        builder.setApiEndpoint(endpoint)
        return this
    }

    fun setTracesApiEndpoint(endpoint: String): HoneycombConfigure {
        builder.setTracesApiEndpoint(endpoint)
        return this
    }

    fun setMetricsApiEndpoint(endpoint: String): HoneycombConfigure {
        builder.setMetricsApiEndpoint(endpoint)
        return this
    }

    fun setLogsApiEndpoint(endpoint: String): HoneycombConfigure {
        builder.setLogsApiEndpoint(endpoint)
        return this
    }

    fun setLogRecordProcessor(logRecordProcessor: LogRecordProcessor): HoneycombConfigure {
        builder.setLogRecordProcessor(logRecordProcessor)
        return this
    }

    fun setSpanProcessor(spanProcessor: SpanProcessor): HoneycombConfigure {
        builder.setSpanProcessor(spanProcessor)
        return this
    }

    fun setSampleRate(sampleRate: Int): HoneycombConfigure {
        builder.setSampleRate(sampleRate)
        return this
    }

    fun setDebug(debug: Boolean): HoneycombConfigure {
        builder.setDebug(debug)
        return this
    }

    fun setServiceName(serviceName: String): HoneycombConfigure {
        builder.setServiceName(serviceName)
        return this
    }

    fun setServiceVersion(appVersion: String): HoneycombConfigure {
        builder.setServiceVersion(appVersion)
        return this
    }

    fun setResourceAttributes(resources: Map<String, String>): HoneycombConfigure {
        builder.setResourceAttributes(resources)
        return this
    }

    fun setHeaders(headers: Map<String, String>): HoneycombConfigure {
        builder.setHeaders(headers)
        return this
    }

    fun setTracesHeaders(headers: Map<String, String>): HoneycombConfigure {
        builder.setTracesHeaders(headers)
        return this
    }

    fun setMetricsHeaders(headers: Map<String, String>): HoneycombConfigure {
        builder.setMetricsHeaders(headers)
        return this
    }

    fun setLogsHeaders(headers: Map<String, String>): HoneycombConfigure {
        builder.setLogsHeaders(headers)
        return this
    }

    fun setTimeout(timeout: Duration): HoneycombConfigure {
        builder.setTimeout(timeout)
        return this
    }

    fun setTracesTimeout(timeout: Duration): HoneycombConfigure {
        builder.setTracesTimeout(timeout)
        return this
    }

    fun setMetricsTimeout(timeout: Duration): HoneycombConfigure {
        builder.setMetricsTimeout(timeout)
        return this
    }

    fun setLogsTimeout(timeout: Duration): HoneycombConfigure {
        builder.setLogsTimeout(timeout)
        return this
    }

    fun setProtocol(protocol: OtlpProtocol): HoneycombConfigure {
        builder.setProtocol(protocol)
        return this
    }

    fun setTracesProtocol(protocol: OtlpProtocol): HoneycombConfigure {
        builder.setTracesProtocol(protocol)
        return this
    }

    fun setMetricsProtocol(protocol: OtlpProtocol): HoneycombConfigure {
        builder.setMetricsProtocol(protocol)
        return this
    }

    fun setLogsProtocol(protocol: OtlpProtocol): HoneycombConfigure {
        builder.setLogsProtocol(protocol)
        return this
    }

    fun setOfflineCachingEnabled(enabled: Boolean): HoneycombConfigure {
        builder.setOfflineCachingEnabled(enabled)
        return this
    }

    internal fun getContext() : Application {
        return context
    }

    internal fun build() : HoneycombOptions {
        return builder.build()
    }
}
